features/restore/src/main/java/ir/mehdiyari/krypt/restore/RestoreViewModel.kt

```diff
@@ -0,0 +1,87 @@
+package ir.mehdiyari.krypt.restore
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import dagger.hilt.android.lifecycle.HiltViewModel
+import ir.mehdiyari.krypt.backup.logic.restore.RestoreKeyGenerator
+import ir.mehdiyari.krypt.backup.logic.restore.RestoreRepository
+import ir.mehdiyari.krypt.cryptography.exceptions.DecryptException
+import ir.mehdiyari.krypt.permission.checkIfAppIsStorageManager
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asSharedFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.launch
+import java.io.FileNotFoundException
+import javax.inject.Inject
+import ir.mehdiyari.krypt.shared.designsystem.resources.R as DesignSystemR
+
+@HiltViewModel
+internal class RestoreViewModel @Inject constructor(
+    private val restoreRepository: RestoreRepository,
+    private val restoreKeyGenerator: RestoreKeyGenerator,
+) : ViewModel() {
+
+    private val _restoreViewState =
+        MutableStateFlow<RestoreViewState>(RestoreViewState.OpenBackupFile)
+    val restoreViewState: StateFlow<RestoreViewState> = _restoreViewState.asStateFlow()
+
+    private val _restoreLoadingState =
+        MutableStateFlow(false)
+    val restoreLoadingState: StateFlow<Boolean> = _restoreLoadingState.asStateFlow()
+
+    private var selectedFilePath: String? = null
+
+    private val _restoreMessageSharedFlow = MutableSharedFlow<Int>()
+    val restoreMessageSharedFlow = _restoreMessageSharedFlow.asSharedFlow()
+
+    private fun restoreBackupFile(
+        filePath: String,
+        password: String,
+    ) {
+        viewModelScope.launch {
```

Comment:
@mehdiyari jan, just for my information, don't we need using `Dispatchers.IO` in this coroutine? why we use `Dispatchers.Main.immediate ` here?

Reply:
This is because of main safety, the dispatcher of this code will be set in the lower layers(use-case, repository or maybe data source). 
You can read more about it here: https://developer.android.com/kotlin/coroutines/coroutines-best-practices#main-safe